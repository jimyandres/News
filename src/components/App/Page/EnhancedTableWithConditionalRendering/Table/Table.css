.table {
  margin: 20px 0;
}
.table-header {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  line-height: 24px;
  font-size: 16px;
  justify-content: space-around;
}

.arrowContainer {
  position: absolute;
  padding-left: 9em;
}
.arrow {
  height: 10px;
  width: 10px;
	border: 1px solid #ef6c00;
	border-width: 2px 2px 0 0;
}
.arrow-up {
  transform: rotate(-45deg);
  margin-top: 7px;
}
.arrow-down {
  transform: rotate(135deg);
  margin-top: -7px;
}
