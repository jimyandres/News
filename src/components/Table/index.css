.table {
  margin: 20px 0;
}

.table-header {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  line-height: 24px;
  font-size: 16px;
  justify-content: space-around;
}

.table-empty {
  margin: 200px;
  text-align: center;
  font-size: 16px;
}

.table-row {
  display: flex;
  line-height: 24px;
  white-space: nowrap;
  margin: 10px 0;
  padding: 10px;
  align-items: center;
  background: #ffffff;
  border: 1px solid #e3e3e3;
}

.table-header > span {
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0 5px
}

.table-row > span {
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0 5px;
}

.arrowContainer {
  position: absolute;
  padding-left: 9em;
}
.arrow {
  height: 10px;
  width: 10px;
	border: 1px solid #ef6c00;
	border-width: 2px 2px 0 0;
}
.arrow-up {
  transform: rotate(-45deg);
  margin-top: 7px;
}
.arrow-down {
  transform: rotate(135deg);
  margin-top: -7px;
}
