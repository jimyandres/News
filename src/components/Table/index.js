import React, { Component } from 'react';
import Button from '../Button';
import { sortBy } from 'lodash';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import FlatButton from 'material-ui/FlatButton';
import './index.css';

const SORTS = {
  NONE: list => list,
  TITLE: list => sortBy(list, 'title'),
  AUTHOR: list => sortBy(list, 'author'),
  COMMENTS: list => sortBy(list, 'num_comments').reverse(),
  POINTS: list => sortBy(list, 'points').reverse(),
};

class Table extends Component {
  constructor(props) {
    super(props);

    this.state = {
      sortKey: 'NONE',
      isSortReverse: false,
    };

    this.onSort = this.onSort.bind(this);
  }

  onSort(sortKey) {
    const isSortReverse = this.state.sortKey === sortKey && !this.state.isSortReverse;
    this.setState({sortKey, isSortReverse});
  }

  render () {
    const {
      list,
      onDismiss,
    } = this.props;

    const {
      sortKey,
      isSortReverse,
    } = this.state;

    const sortedList = SORTS[sortKey](list);
    const reverseSortedList = isSortReverse
      ? sortedList.reverse()
      : sortedList;

    return (
      <div className="table">
        <div className="table-header">
          <SortList
            sortKeys={[
              {name: 'Title', className:largeColumn},
              {name: 'Author', className:midColumn},
              {name: 'Comments', className:smallColumn},
              {name: 'Points', className:smallColumn},
            ]}
            onSort={this.onSort}
            activeSortKey={sortKey}
            isSortReverse={isSortReverse}
          />
          <FlatButton primary style={smallColumn}>
            Archive
          </FlatButton>
        </div>
        { reverseSortedList.map(item =>
          <div key={item.objectID} className="table-row">
            <span style={largeColumn}>
              <a href={item.url}>{ item.title }</a>
            </span>
            <span style={midColumn}>{item.author}</span>
            <span style={smallColumn}>{item.num_comments}</span>
            <span style={smallColumn}>{item.points}</span>
            <span style={smallColumn}>
              <Button
                onClick={() => onDismiss(item.objectID)}
                className="button-inline"
              >
                Dismiss
              </Button>
            </span>
          </div>
        )}
      </div>
    );
  }
}

Table.propTypes = {
  list: PropTypes.array,
  sortKey: PropTypes.string,
  onSort: PropTypes.func,
  onDismiss: PropTypes.func,
};

const SortList = ({sortKeys, ...rest}) => {
  return sortKeys.map(key =>
    <span key={key.name} style={key.className}>
      <Sort
        sortKey={key.name.toUpperCase()}
        {... rest}
      >
        {key.name}
      </Sort>
    </span>
  );
}

const Sort = ({sortKey, onSort, activeSortKey, isSortReverse, children}) => {
  const sortClass = classNames(
    'button-inline',
    {'button-active': sortKey === activeSortKey}
  );
  const arrowSort = classNames(
    {'arrow arrow-up': sortKey === activeSortKey && !isSortReverse},
    {'arrow arrow-down': sortKey === activeSortKey && isSortReverse}
  );
  return (
    <Button
      onClick={() => onSort(sortKey)}
      className={sortClass}
    >
      {children}<div className={arrowSort}></div>
    </Button>
  );
}


const largeColumn = {
  width: '40%',
};
const midColumn = {
  width: '30%',
};
const smallColumn = {
  width: '10%',
};

export default Table;
