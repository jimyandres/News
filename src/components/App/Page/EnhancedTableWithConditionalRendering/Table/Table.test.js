import React from 'react';
import ReactDOM from 'react-dom';
import { configure, mount, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallowToJson } from 'enzyme-to-json';
import Table from './index';

configure({ adapter: new Adapter() });

describe('Table', () => {

	const props = {
		list: [
			{
				'title': 'Things to learn in React before using Redux',
				'url': 'https://www.robinwieruch.de/learn-react-before-using-redux/',
				'author': 'callumlocke',
				'points': 370,
				'num_comments': 104,
				'_tags': [
					'story',
					'author_callumlocke',
					'story_14811577'
				],
				'objectID': '14811577',
			},
			{
				'title': 'Show HN: ORY Editor â€“ A rich editor for the browser, built with React and Redux',
				'url': 'https://github.com/ory/editor?branch=master',
				'author': 'jswizzard',
				'points': 369,
				'num_comments': 98,
				'_tags': [
					'story',
					'author_jswizzard',
					'story_14417340',
					'show_hn'
				],
				'objectID': '14417340',
			},
			{
				'title': 'A SoundCloud client in React and Redux',
				'url': 'http://www.robinwieruch.de/the-soundcloud-client-in-react-redux/',
				'author': 'rwieruch',
				'points': 253,
				'num_comments': 69,
				'_tags': [
					'story',
					'author_rwieruch',
					'story_11890229'
				],
				'objectID': '11890229',
			},
			{
				'title': 'Systemd redux: The end of Linux',
				'url': 'http://blog.lusis.org/blog/2014/11/20/systemd-redux/',
				'author': 'bcantrill',
				'points': 235,
				'num_comments': 463,
				'_tags': [
					'story',
					'author_bcantrill',
					'story_8639317'
				],
				'objectID': '8639317',
			}
		]
	};

	it('renders without crashing', () => {
		const div = document.createElement('div');
		ReactDOM.render(<Table { ...props } />, div);
	});

	test('has a valid snapshot', () => {
		const component = render(<Table { ...props } />);
		expect(shallowToJson(component)).toMatchSnapshot();
	});

	it('shows two items in list', () => {
		const element = mount(<Table { ...props } />);
		expect(element.find('ExpansionPanel').length).toBe(4);
	});

});
