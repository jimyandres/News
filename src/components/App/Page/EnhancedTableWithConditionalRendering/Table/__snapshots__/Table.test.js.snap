// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table has a valid snapshot 1`] = `null`;
